polytracker/include/taintdag/fdmapping.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef POLYTRACKER_TAINTDAG_FDMAPPING_H
+#define POLYTRACKER_TAINTDAG_FDMAPPING_H
+
+#include <mutex>
+#include <optional>
+#include <string_view>
+
+#include "taintdag/error.hpp"
+#include "taintdag/taint.hpp"
+
+namespace taintdag {
+
+
+  // TODO (hbrodin): There might be a better way to do this... The assumption is that very few files will be added...
+  // TODO (hbrodin): Probably need to consider alignment issues for some of the types.
+  class FDMapping {
+
+    public:
+
+      using length_type = uint32_t;
+      using fd_type = int;
+
+      const size_t length_offset = sizeof(fd_type);
+      const size_t string_offset = length_offset + sizeof(length_type);
+
+      // Following each FDMappingHdr is namelen bytes of name string
+      //  if prealloc_begin == 0 it means no preallocation.
+      // TODO (hbrodin): Consider alignment and padding of this structure.
+      struct FDMappingHdr {
+        fd_type fd;
+        length_type namelen;
+        label_t prealloc_begin;
+        label_t prealloc_end;
+      };
+
+      FDMapping(char *begin, char *end) 
+       : begin_{begin}, end_{end} {
+      }
+
+      size_t get_size() const {
+        std::unique_lock l{m_};
+        return offset_;
+      }
+
+      // The only reason this could fail is if there is no space left (either storage or source_index)
+      std::optional<source_index_t> add_mapping(int fd, std::string_view name, std::optional<taint_range_t> preallocated_labels= {}) {
```

Comment:
Consider keeping the mappings separate from the filenames, to be able to directly index into the mappings. Add a name offset field into each header where to locate the filename.